include/minisketch.h

```diff
@@ -151,6 +151,156 @@ size_t minisketch_compute_max_elements(uint32_t bits, size_t capacity, uint32_t
+    static Minisketch CreateFP(uint32_t bits, uint32_t implementation, size_t max_elements, uint32_t fpbits) noexcept
+    {
+        return Minisketch(bits, implementation, ComputeCapacity(bits, max_elements, fpbits));
+    }
+
+    /** See minisketch_get_bits(). */
+    uint32_t GetBits() const noexcept { return minisketch_bits(m_minisketch.get()); }
+
+    /** See minisketch_get_capacity(). */
+    size_t GetCapacity() const noexcept { return minisketch_capacity(m_minisketch.get()); }
+
+    /** See minisketch_get_implementation(). */
+    uint32_t GetImplementation() const noexcept { return minisketch_implementation(m_minisketch.get()); }
+
+    /** See minisketch_set_seed(). */
+    Minisketch& SetSeed(uint64_t seed) noexcept
+    {
+        minisketch_set_seed(m_minisketch.get(), seed);
+        return *this;
+    }
+
+    /** See miniksetch_add_element(). */
+    Minisketch& Add(uint64_t element) noexcept
+    {
+        minisketch_add_uint64(m_minisketch.get(), element);
+        return *this;
+    }
+
+    /** See minisketch_merge(). */
+    Minisketch& Merge(const Minisketch& sketch) noexcept
+    {
+        minisketch_merge(m_minisketch.get(), sketch.m_minisketch.get());
+        return *this;
+    }
+
+    /** Decode this sketch into the result vector, which must have a size equal to the maximum allowed elements. */
+    bool Decode(std::vector<uint64_t>& result) const noexcept
+    {
+        ssize_t ret = minisketch_decode(m_minisketch.get(), result.size(), result.data());
+        if (ret == -1) return false;
+        result.resize(ret);
+        return true;
+    }
+
+    /** See minisketch_serialized_size(). */
+    size_t GetSerializedSize() const noexcept { return minisketch_serialized_size(m_minisketch.get()); }
+
+    /** Serialize the sketch as a byte vector. */
+    std::vector<unsigned char> Serialize() const noexcept
+    {
+        std::vector<unsigned char> result(GetSerializedSize());
+        minisketch_serialize(m_minisketch.get(), result.data());
+        return result;
+    }
+
+    /** Deserialize into this sketch from an object containing its bytes (which has data() and size() members). */
+    template<typename T>
+    Minisketch& Deserialize(
+        const T& obj,
+        typename std::enable_if<
+            std::is_convertible<typename std::remove_pointer<decltype(obj.data())>::type (*)[], const unsigned char (*)[]>::value &&
+            std::is_convertible<decltype(obj.size()), std::size_t>::value,
+            std::nullptr_t
+        >::type = nullptr) noexcept
+    {
+        assert(GetSerializedSize() == obj.size());
+        minisketch_deserialize(m_minisketch.get(), obj.data());
+        return *this;
+    }
+
+#if __cplusplus >= 201703L
+    /** C++17 only: decode with a specified number of max elements into an optional vector. */
+    std::optional<std::vector<uint64_t>> Decode(size_t max_elements) const noexcept
+    {
+        std::vector<uint64_t> result(max_elements);
```

Comment:
How is this noexcept? It's constructing a vector using (3) here: https://en.cppreference.com/w/cpp/container/vector/vector which is not noexcept.

In general, my understanding was that any function that can allocate on the heap shouldn't be noexcept.

Reply:
Yes and no. You can make something `noexcept` even if it's going to throw, if you're ok with the program terminating when it does. I think that's... mostly true here; I personally think OOM is a perfectly fine reason to terminate a program, as the situations that trigger it generally make it pretty much impossible to continue anyway.

Still, I've removed it here. Handling OOM isn't entirely consistent in the codebase currently; in some places it's caught, and in others it isn't. Let's leave addressing that for another PR.